# 20MAI0005_Abhishek_Kumar

# Introduction to NumPy


The learning objectives of this section are:

* Understand advantages of vectorised code using NumPy (over standard python ways)
* Create NumPy arrays
    * Convert lists and tuples to NumPy arrays 
    * Create (initialise) arrays
* Inspect the structure and content of arrays
* Subset, slice, index and iterate through arrays
* Compare computation times in NumPy and standard Python lists

### NumPy Basics

NumPy is a library written for scientific computing and data analysis. It stands for numerical python.

The most basic object in NumPy is the ```ndarray```, or simply an ```array```, which is an **n-dimensional, homogenous** array. By homogenous, we mean that all the elements in a NumPy array have to be of the **same data type**, which is commonly numeric (float or integer). 

Let's see some examples of arrays.

In [2]:
# Import the numpy library
# np is simply an alias, you may use any other alias, though np is quite standard
import numpy as np

In [3]:
# Creating a 1-D array using a list
# np.array() takes in a list or a tuple as argument, and converts into an array
array_1d = np.array([2, 4, 5, 6, 7, 9])
print(array_1d)
print(type(array_1d))

[2 4 5 6 7 9]
<class 'numpy.ndarray'>


In [4]:
# Creating a 2-D array using two lists
array_2d = np.array([[2, 3, 4], [5, 8, 7]])
print(array_2d)

[[2 3 4]
 [5 8 7]]


### Advantages of NumPy 

What is the use of arrays over lists, specifically for data analysis? Putting crudely, it is **convenience and speed **:<br>
1. You can write **vectorised** code on numpy arrays, not on lists, which is **convenient to read and write, and concise**. 
2. Numpy is **much faster** than the standard python ways to do computations.

Vectorised code typically does not contain explicit looping and indexing etc. (all of this happens behind the scenes, in precompiled C-code), and thus it is much more concise.

Let's see an example of convenience, we'll see one later for speed. 

In [5]:
list_1 = [3, 6, 7, 5]
list_2 = [4, 5, 1, 7]

# the list way to do it: map a function to the two lists
product_list = list(map(lambda x, y: x*y, list_1, list_2))
print(product_list)


[12, 30, 7, 35]


In [6]:
# The numpy array way to do it: simply multiply the two arrays
array_1 = np.array(list_1)
array_2 = np.array(list_2)

array_3 = array_1*array_2
print(array_3)
print(type(array_3))

[12 30  7 35]
<class 'numpy.ndarray'>


As you can see, the NumPy way is clearly more concise.

Even simple mathematical operations on lists require for loops, unlike with arrays. For example, to calculate the square of every number in a list:

In [8]:
# Square a list
list_squared = [i**2 for i in list_1]

# Square a numpy array
array_squared = array_1**2

print(list_squared)
print(array_squared)

[9, 36, 49, 25]
[ 9 36 49 25]


This was with 1-D arrays. You'll often work with 2-D arrays (matrices), where the difference would be even greater. With lists, you'll have to store matrices as lists of lists and loop through them. With NumPy, you simply multiply the matrices.

### Creating NumPy Arrays 

There are multiple ways to create numpy arrays, the most commmon ones being:
* Convert lists or tuples to arrays using ```np.array()```, as done above
* Initialise arrays of fixed size (when the size is known) 


In [10]:
# Convert lists or tuples to arrays using np.array()
# Note that np.array(2, 5, 6, 7) will throw an error - you need to pass a list or a tuple
array_from_list = np.array([2, 5, 6, 7]) 
array_from_tuple = np.array((4, 5, 8, 9))

print(array_from_list)
print(array_from_tuple)

[2 5 6 7]
[4 5 8 9]


The other common way is to initialise arrays. You do this when you know the size of the array beforehand.

The following ways are commonly used:
* ```np.ones()```: Create array of 1s
* ```np.zeros()```: Create array of 0s
* ```np.random.random()```: Create array of random numbers
* ```np.arange()```: Create array with increments of a fixed step size
* ```np.linspace()```: Create array of fixed length

In [11]:
# Tip: Use help to see the syntax when required
help(np.ones)

Help on function ones in module numpy:

ones(shape, dtype=None, order='C')
    Return a new array of given shape and type, filled with ones.
    
    Parameters
    ----------
    shape : int or sequence of ints
        Shape of the new array, e.g., ``(2, 3)`` or ``2``.
    dtype : data-type, optional
        The desired data-type for the array, e.g., `numpy.int8`.  Default is
        `numpy.float64`.
    order : {'C', 'F'}, optional, default: C
        Whether to store multi-dimensional data in row-major
        (C-style) or column-major (Fortran-style) order in
        memory.
    
    Returns
    -------
    out : ndarray
        Array of ones with the given shape, dtype, and order.
    
    See Also
    --------
    ones_like : Return an array of ones with shape and type of input.
    empty : Return a new uninitialized array.
    zeros : Return a new array setting values to zero.
    full : Return a new array of given shape filled with value.
    
    
    Examples
    --------
   

In [12]:
# Creating a 5 x 3 array of ones
np.ones((5, 3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [13]:
# Notice that, by default, numpy creates data type = float64
# Can provide dtype explicitly using dtype
np.ones((5, 3), dtype = np.int)

array([[1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1],
       [1, 1, 1]])

In [14]:
# Creating array of zeros
np.zeros(4, dtype = np.int)

array([0, 0, 0, 0])

In [15]:
# Array of random numbers
np.random.random([3, 4])

array([[0.5394394 , 0.69362432, 0.59027808, 0.80269875],
       [0.1493517 , 0.31421957, 0.31463597, 0.09240781],
       [0.91650201, 0.43978087, 0.78037313, 0.57737993]])

In [16]:
# np.arange()
# np.arange() is the numpy equivalent of range()
# Notice that 10 is included, 100 is not, as in standard python lists

# From 10 to 100 with a step of 5
numbers = np.arange(10, 100, 5)
print(numbers)

[10 15 20 25 30 35 40 45 50 55 60 65 70 75 80 85 90 95]


In [17]:
# np.linspace()
# Sometimes, you know the length of the array, not the step size

# Array of length 25 between 15 and 18
np.linspace(15, 18, 25)

array([15.   , 15.125, 15.25 , 15.375, 15.5  , 15.625, 15.75 , 15.875,
       16.   , 16.125, 16.25 , 16.375, 16.5  , 16.625, 16.75 , 16.875,
       17.   , 17.125, 17.25 , 17.375, 17.5  , 17.625, 17.75 , 17.875,
       18.   ])

Apart from the methods mentioned above, there are a few more NumPy functions that you can use to create special NumPy arrays:

-  `np.full()`: Create a constant array of any number ‘n’
-  `np.tile()`: Create a new array by repeating an existing array for a particular number of times
-  `np.eye()`: Create an identity matrix of any dimension
-  `np.random.randint()`: Create a random array of integers within a particular range

In this case, numpy is **an order of magnitude faster** than lists. This is with arrays of size in millions, but you may work on much larger arrays of sizes in order of billions. Then, the difference is even larger.

Some reasons for such difference in speed are:
* NumPy is written in C, which is basically being executed behind the scenes
* NumPy arrays are more compact than lists, i.e. they take much lesser storage space than lists

# Pandas


Pandas is a library built using NumPy specifically for data analysis. You'll be using Pandas heavily for data manipulation, visualisation, building machine learning models, etc. 

There are two main data structures in Pandas - Series and Dataframes. The default way to store data is dataframes, and thus manipulating dataframes quickly is probably the most important skill set for data analysis. 

### 1. The Pandas Series 

A series is similar to a 1-D numpy array, and contains scalar values of the same type (numeric, character, datetime etc.). 
A dataframe is simply a table where each column is a pandas series.


#### Creating Pandas Series

Series are one-dimensional array-like structures, though unlike numpy arrays, they often contain non-numeric data (characters, dates, time, booleans etc.)

You can create pandas series from array-like objects using ```pd.Series()```.

In [1]:
# import pandas, pd is an alias
import pandas as pd
import numpy as np

# Creating a numeric pandas series
s = pd.Series([2, 4, 5, 6, 9])
print(s)
print(type(s))

0    2
1    4
2    5
3    6
4    9
dtype: int64
<class 'pandas.core.series.Series'>


Note that each element in the Series has an index, and the index starts at 0 as usual.

In [2]:
# creating a series of characters
# notice that the 'dtype' here is 'object'
char_series = pd.Series(['a', 'b', 'af'])
char_series

0     a
1     b
2    af
dtype: object

In [3]:
# creating a series of type datetime
date_series = pd.date_range(start = '11-09-2017', end = '12-12-2017')
date_series
type(date_series)


pandas.core.indexes.datetimes.DatetimeIndex

#### Indexing Series

Indexing series is exactly same as 1-D numpy arrays - index starts at 0.

In [4]:
# Indexing pandas series: Same as indexing 1-d numpy arrays or lists
# accessing the fourth element
s[3]

# accessing elements starting index = 2 till the end
s[2:]

2    5
3    6
4    9
dtype: int64

In [5]:
# accessing the second and the fourth elements
# note that s[1, 3] will not work, you need to pass the indices [1, 3] as a list inside the original []
s[[1, 3]]

1    4
3    6
dtype: int64

#### Explicitly specifying indices

You might have noticed that while creating a series, Pandas automatically indexes it from 0 to (n-1), n being the number of rows. But if we want, we can also explicitly set the index ourselves, using the ‘index’ argument while creating the series using `pd.Series()`

In [6]:
# Indexing explicitly
pd.Series([0, 1, 2], index = ['a', 'b', 'c'])

a    0
b    1
c    2
dtype: int64

In [7]:
# You can also give the index as a sequence or use functions to specify the index
# But always make sure that the number of elements in the index list is equal to the number of elements specified in the series

pd.Series(np.array(range(0,10))**2, index = range(0,10))

0     0
1     1
2     4
3     9
4    16
5    25
6    36
7    49
8    64
9    81
dtype: int32

Usually, you will work with Series only as a part of dataframes. Let's study the basics of dataframes.

### The Pandas Dataframe

Dataframe is the most widely used data-structure in data analysis. It is a table with rows and columns, with rows having an index and columns having meaningful names.

#### Creating dataframes from dictionaries

There are various ways of creating dataframes, such as creating them from dictionaries, JSON objects, reading from txt, CSV files, etc. 

In [9]:
# keys become column names
df = pd.DataFrame({'name': ['Vinay', 'Kushal', 'Aman', 'Saif'], 
                   'age': [22, 25, 24, 28], 
                    'occupation': ['engineer', 'doctor', 'data analyst', 'teacher']})
df

name  age    occupation
0   Vinay   22      engineer
1  Kushal   25        doctor
2    Aman   24  data analyst
3    Saif   28       teacher

#### Importing CSV data files as pandas dataframes 

For the upcoming exercises, we will use a dataset of a retail store having details about the orders placed, customers, product details, sales, profits etc. 



In [10]:
# reading a CSV file as a dataframe
market_df = pd.read_csv("global_sales_data/market_fact.csv")

Usually, dataframes are imported as CSV files, but sometimes it is more convenient to convert dictionaries 
into dataframes. For e.g. when the raw data is in a JSON format (which is not uncommon), you can easily convert it into a dictionary, and then into a dataframe. 

You will learn how to convert JSON objects to dataframes later.

#### Reading and Summarising Dataframes

After you import a dataframe, you'd want to quickly understand its structure, shape, meanings of rows and columns etc. Further, you may want to look at summary statistics - such as mean, percentiles etc.

In [11]:
# Looking at the top and bottom entries of dataframes
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

In [12]:
market_df.tail()

Ord_id  Prod_id   Ship_id    Cust_id      Sales  Discount  \
8394  Ord_5353   Prod_4  SHP_7479  Cust_1798  2841.4395      0.08   
8395  Ord_5411   Prod_6  SHP_7555  Cust_1798   127.1600      0.10   
8396  Ord_5388   Prod_6  SHP_7524  Cust_1798   243.0500      0.02   
8397  Ord_5348  Prod_15  SHP_7469  Cust_1798  3872.8700      0.03   
8398  Ord_5459   Prod_6  SHP_7628  Cust_1798   603.6900      0.00   

      Order_Quantity  Profit  Shipping_Cost  Product_Base_Margin  
8394              28  374.63           7.69                 0.59  
8395              20  -74.03           6.92                 0.37  
8396              39  -70.85           5.35                 0.40  
8397              23  565.34          30.00                 0.62  
8398              47  131.39           4.86                 0.38

Here, each row represents an order placed at a retail store. Notice the index associated with each row - starts at 0 and ends at 8398, implying that there were 8399 orders placed.

In [15]:
# Looking at the datatypes of each column
market_df.info()

# Note that each column is basically a pandas Series of length 8399
# The ID columns are 'objects', i.e. they are being read as characters
# The rest are numeric (floats or int)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Ord_id               8399 non-null   object 
 1   Prod_id              8399 non-null   object 
 2   Ship_id              8399 non-null   object 
 3   Cust_id              8399 non-null   object 
 4   Sales                8399 non-null   float64
 5   Discount             8399 non-null   float64
 6   Order_Quantity       8399 non-null   int64  
 7   Profit               8399 non-null   float64
 8   Shipping_Cost        8399 non-null   float64
 9   Product_Base_Margin  8336 non-null   float64
dtypes: float64(5), int64(1), object(4)
memory usage: 656.3+ KB


In [16]:
# Describe gives you a summary of all the numeric columns in the dataset
market_df.describe()

Sales     Discount  Order_Quantity        Profit  Shipping_Cost  \
count   8399.000000  8399.000000     8399.000000   8399.000000    8399.000000   
mean    1775.878179     0.049671       25.571735    181.184424      12.838557   
std     3585.050525     0.031823       14.481071   1196.653371      17.264052   
min        2.240000     0.000000        1.000000 -14140.700000       0.490000   
25%      143.195000     0.020000       13.000000    -83.315000       3.300000   
50%      449.420000     0.050000       26.000000     -1.500000       6.070000   
75%     1709.320000     0.080000       38.000000    162.750000      13.990000   
max    89061.050000     0.250000       50.000000  27220.690000     164.730000   

       Product_Base_Margin  
count          8336.000000  
mean              0.512513  
std               0.135589  
min               0.350000  
25%               0.380000  
50%               0.520000  
75%               0.590000  
max               0.850000

In [17]:
# Column names
market_df.columns

Index(['Ord_id', 'Prod_id', 'Ship_id', 'Cust_id', 'Sales', 'Discount',
       'Order_Quantity', 'Profit', 'Shipping_Cost', 'Product_Base_Margin'],
      dtype='object')

In [18]:
# The number of rows and columns
market_df.shape

(8399, 10)

In [19]:
# You can extract the values of a dataframe as a numpy array using df.values 
market_df.values

array([['Ord_5446', 'Prod_16', 'SHP_7609', ..., -30.51, 3.6, 0.56],
       ['Ord_5406', 'Prod_13', 'SHP_7549', ..., 4.56, 0.93, 0.54],
       ['Ord_5446', 'Prod_4', 'SHP_7610', ..., 1148.9, 2.5, 0.59],
       ...,
       ['Ord_5388', 'Prod_6', 'SHP_7524', ..., -70.85, 5.35, 0.4],
       ['Ord_5348', 'Prod_15', 'SHP_7469', ..., 565.34, 30.0, 0.62],
       ['Ord_5459', 'Prod_6', 'SHP_7628', ..., 131.39, 4.86, 0.38]],
      dtype=object)

#### Indices 

An important concept in pandas dataframes is that of *row indices*. By default, each row is assigned indices starting from 0, and are represented at the left side of the dataframe. 

In [20]:
market_df.head()

Ord_id  Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
0  Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
1  Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
2  Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
3  Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
4  Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

    Profit  Shipping_Cost  Product_Base_Margin  
0   -30.51           3.60                 0.56  
1     4.56           0.93                 0.54  
2  1148.90           2.50                 0.59  
3   729.34          14.30                 0.37  
4  1219.87          26.30                 0.38

Now, arbitrary numeric indices are difficult to read and work with. Thus, you may want to change the indices of the df to something more meanigful.

Let's change the index to Ord_id (unique id of each order), so that you can select rows using the order ids directly.

In [22]:
# Setting index to Ord_id
market_df.set_index('Ord_id', inplace = True)
market_df.head()

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_5446  Prod_16  SHP_7609  Cust_1818   136.81      0.01              23   
Ord_5406  Prod_13  SHP_7549  Cust_1818    42.27      0.01              13   
Ord_5446   Prod_4  SHP_7610  Cust_1818  4701.69      0.00              26   
Ord_5456   Prod_6  SHP_7625  Cust_1818  2337.89      0.09              43   
Ord_5485  Prod_17  SHP_7664  Cust_1818  4233.15      0.08              35   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_5446   -30.51           3.60                 0.56  
Ord_5406     4.56           0.93                 0.54  
Ord_5446  1148.90           2.50                 0.59  
Ord_5456   729.34          14.30                 0.37  
Ord_5485  1219.87          26.30                 0.38

Having meaningful row labels as indices helps you to select (subset) dataframes easily. You will study selecting dataframes in the next section. 

#### Sorting dataframes

You can sort dataframes in two ways - 1) by the indices and 2) by the values.  


In [23]:
# Sorting by index
# axis = 0 indicates that you want to sort rows (use axis=1 for columns)
market_df.sort_index(axis = 0, ascending = False)

Prod_id   Ship_id   Cust_id     Sales  Discount  Order_Quantity  \
Ord_id                                                                      
Ord_999   Prod_15  SHP_1383  Cust_361   5661.08      0.00              33   
Ord_998    Prod_8  SHP_1380  Cust_372    750.66      0.00              33   
Ord_998    Prod_5  SHP_1382  Cust_372   2149.37      0.03              42   
Ord_998    Prod_8  SHP_1381  Cust_372    254.32      0.01               8   
Ord_997   Prod_14  SHP_1379  Cust_365  28761.52      0.04               8   
...           ...       ...       ...       ...       ...             ...   
Ord_1001   Prod_5  SHP_1385  Cust_374   1981.26      0.07              49   
Ord_1000   Prod_6  SHP_1384  Cust_373    334.71      0.01              25   
Ord_100    Prod_8   SHP_138   Cust_58    121.12      0.10               3   
Ord_10     Prod_3    SHP_13   Cust_10     80.61      0.02              15   
Ord_1      Prod_1     SHP_1    Cust_1    261.54      0.04               6   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_999   1055.47          30.00                 0.62  
Ord_998    120.05           4.00                 0.60  
Ord_998    217.87          19.99                 0.55  
Ord_998   -117.39           6.50                 0.79  
Ord_997    285.11          24.49                 0.37  
...           ...            ...                  ...  
Ord_1001   100.80           8.66                 0.76  
Ord_1000    31.74           6.47                 0.38  
Ord_100   -118.82           1.99                 0.44  
Ord_10      -4.72           2.99                 0.37  
Ord_1     -213.25          35.00                 0.80  

[8399 rows x 9 columns]

In [24]:
# Sorting by values

# Sorting in increasing order of Sales
market_df.sort_values(by='Sales').head()

Prod_id   Ship_id    Cust_id  Sales  Discount  Order_Quantity  \
Ord_id                                                                    
Ord_704    Prod_7   SHP_964   Cust_242   2.24      0.01               1   
Ord_149    Prod_3  SHP_7028  Cust_1712   3.20      0.09               1   
Ord_4270   Prod_7  SHP_5959  Cust_1450   3.23      0.06               2   
Ord_4755  Prod_13  SHP_6628  Cust_1579   3.41      0.06               1   
Ord_2252   Prod_3  SHP_3064   Cust_881   3.42      0.05               1   

          Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                
Ord_704    -1.97           0.70                 0.37  
Ord_149    -3.16           1.49                 0.37  
Ord_4270   -2.73           0.70                 0.81  
Ord_4755   -1.78           0.70                 0.56  
Ord_2252   -2.91           1.49                 0.37

In [25]:
# Sorting in decreasing order of Shipping_Cost
market_df.sort_values(by='Shipping_Cost', ascending = False).head()

Prod_id   Ship_id   Cust_id      Sales  Discount  Order_Quantity  \
Ord_id                                                                       
Ord_1751  Prod_15  SHP_2426  Cust_597  14740.510      0.00              46   
Ord_839   Prod_11  SHP_1361  Cust_364  12689.870      0.04              44   
Ord_1741  Prod_11  SHP_2411  Cust_595  15168.820      0.02              26   
Ord_417   Prod_11   SHP_561  Cust_156  20333.816      0.02              45   
Ord_1581  Prod_15  SHP_2184  Cust_519   2573.920      0.07              17   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_1751  3407.73         164.73                 0.56  
Ord_839   -169.23         154.12                 0.76  
Ord_1741 -1096.78         147.12                 0.80  
Ord_417  -1430.45         147.12                 0.80  
Ord_1581   117.23         143.71                 0.55

In [26]:
# Sorting by more than two columns

# Sorting in ascending order of Sales for each Product
market_df.sort_values(by=['Prod_id', 'Sales'], ascending = False)

Prod_id   Ship_id    Cust_id    Sales  Discount  Order_Quantity  \
Ord_id                                                                     
Ord_2197  Prod_9  SHP_2994   Cust_827  7522.80      0.04              48   
Ord_4356  Prod_9  SHP_6074  Cust_1481  6831.72      0.01              41   
Ord_262   Prod_9   SHP_358    Cust_66  6553.45      0.03              39   
Ord_4059  Prod_9  SHP_5660  Cust_1378  5587.20      0.05              36   
Ord_2973  Prod_9  SHP_6073  Cust_1480  5410.95      0.09              36   
...          ...       ...        ...      ...       ...             ...   
Ord_3713  Prod_1  SHP_5145  Cust_1307    27.83      0.09               2   
Ord_2746  Prod_1  SHP_3767  Cust_1030    22.61      0.03               1   
Ord_439   Prod_1   SHP_587   Cust_136    18.73      0.05               1   
Ord_2314  Prod_1  SHP_3171   Cust_899    18.16      0.03               1   
Ord_286   Prod_1   SHP_387    Cust_90    18.15      0.04               1   

           Profit  Shipping_Cost  Product_Base_Margin  
Ord_id                                                 
Ord_2197  3187.37          19.99                 0.39  
Ord_4356  3081.02          19.99                 0.39  
Ord_262   2969.81          19.99                 0.39  
Ord_4059  2254.16          19.99                 0.39  
Ord_2973  2077.91          19.99                 0.39  
...           ...            ...                  ...  
Ord_3713   -22.14           9.45                 0.60  
Ord_2746    -8.40           7.51                 0.57  
Ord_439     -6.68           6.13                 0.57  
Ord_2314    -7.25           6.13                 0.57  
Ord_286     -7.26           6.13                 0.57  

[8399 rows x 9 columns]